src/renderer/dx/DxFontInfo.h

```diff
@@ -0,0 +1,98 @@
+// Licensed under the MIT license.
+
+#pragma once
+
+#include <dwrite.h>
+#include <dwrite_1.h>
+#include <dwrite_2.h>
+#include <dwrite_3.h>
+
+namespace Microsoft::Console::Render
+{
+    class DxFontInfo
+    {
+    public:
+        DxFontInfo() noexcept;
+
+        DxFontInfo(std::wstring_view familyName,
+                   unsigned int weight,
+                   DWRITE_FONT_STYLE style,
+                   DWRITE_FONT_STRETCH stretch);
+
+        DxFontInfo(std::wstring_view familyName,
+                   DWRITE_FONT_WEIGHT weight,
+                   DWRITE_FONT_STYLE style,
+                   DWRITE_FONT_STRETCH stretch);
+
+        bool operator==(const DxFontInfo& other) const noexcept;
+
+        std::wstring_view GetFamilyName() const noexcept;
+        void SetFamilyName(const std::wstring_view familyName);
+
+        DWRITE_FONT_WEIGHT GetWeight() const noexcept;
+        void SetWeight(const DWRITE_FONT_WEIGHT weight) noexcept;
+
+        DWRITE_FONT_STYLE GetStyle() const noexcept;
+        void SetStyle(const DWRITE_FONT_STYLE style) noexcept;
+
+        DWRITE_FONT_STRETCH GetStretch() const noexcept;
+        void SetStretch(const DWRITE_FONT_STRETCH stretch) noexcept;
+
+        bool GetFallback() const noexcept;
+
+        void SetFromEngine(const std::wstring_view familyName,
+                           const DWRITE_FONT_WEIGHT weight,
+                           const DWRITE_FONT_STYLE style,
+                           const DWRITE_FONT_STRETCH stretch);
+
+        [[nodiscard]] ::Microsoft::WRL::ComPtr<IDWriteFontFace1> ResolveFontFaceWithFallback(gsl::not_null<IDWriteFactory1*> dwriteFactory,
+                                                                                             std::wstring& localeName);
+
+    private:
+        [[nodiscard]] ::Microsoft::WRL::ComPtr<IDWriteFontFace1> _FindFontFace(gsl::not_null<IDWriteFactory1*> dwriteFactory, std::wstring& localeName);
+
+        [[nodiscard]] std::wstring _GetFontFamilyName(gsl::not_null<IDWriteFontFamily*> const fontFamily,
+                                                      std::wstring& localeName);
+
+        [[nodiscard]] const Microsoft::WRL::ComPtr<IDWriteFontCollection1>& _NearbyCollection(gsl::not_null<IDWriteFactory1*> dwriteFactory) const;
+
+        [[nodiscard]] static std::vector<std::filesystem::path> s_GetNearbyFonts();
+
+        mutable ::Microsoft::WRL::ComPtr<IDWriteFontCollection1> _nearbyCollection;
+
+        // The font name we should be looking for
+        std::wstring _familyName;
+
+        // The weight (bold, light, etc.)
+        DWRITE_FONT_WEIGHT _weight;
+
+        // Normal, italic, etc.
+        DWRITE_FONT_STYLE _style;
+
+        // The stretch of the font is the spacing between each letter
+        DWRITE_FONT_STRETCH _stretch;
+
+        // Indicates whether we couldn't match the user request and had to choose from a hardcoded default list.
+        bool _didFallback;
+    };
+
+    struct DxFontInfoHash

```

Comment:
Normally you specialize `std::hash<DxFontInfo>` instead.

Reply:
I’m a bit against the idea of arbitrarily injecting things into ::std namespace. So I choose the other way. If you guys all favor the first way, I’m happy to change it.

Reply:
[`std::hash`](https://en.cppreference.com/w/cpp/utility/hash) specifically is an extension point at which user code is supposed to be injected via specialisation, so that hashing, hash combining, putting into `std::unordered_map` and workalikes, [`std::hash<std::variant>`](https://en.cppreference.com/w/cpp/utility/variant/hash), and any other operation that assumes `std::hash`, "just works" without needing to _also_ override the reference to `std::hash`.

Particularly, if you implement a hash function for a structure containing a `DxFontInfo`, it should look like the hash function here, combining `std::hash` of the members, rather than needing to handle `DxFontInfoHash` specially, an _assuming_ from the name that implements [`Hash`](https://en.cppreference.com/w/cpp/named_req/Hash).

Reply:
Thanks guys. I learned a lot from this thread. I'm gonna leave the choice to Captain Dustin. When he gets to this PR and make a decision, I'll be happy to use either of the above ways.

Reply:
My personal preference is that we properly extend `std::hash`; it means that we can use the STL containers with less customization, and perhaps that leads to a higher likelihood that they will be deduplicated as "identical code" :smile: